### PR TITLE
fix(vector-store): reconstruct FAISS vectors

### LINKED ISSUE #844: FAISSIndex.get_vector() hardcoded to return None, preventing raw vector retrieval
## Summary

`FAISSIndex.get_vector()` is currently hardcoded to return `None`, making it impossible to retrieve stored embeddings even when the underlying FAISS index supports reconstruction.

## Description

After storing vectors in the FAISS backend, calling `get_vector()` never returns the original embedding because the implementation is currently:

```python
# Note: FAISS doesn't directly support retrieval by index in all cases
# This is a simplified approach
return None
```

This silently drops the capability of retrieving raw vectors.

## Root Cause

While some FAISS index types (such as PQ or IVFPQ) cannot perfectly reconstruct stored vectors, the default index used by the wrapper is `IndexFlatL2`, which fully supports exact reconstruction.

The current implementation ignores this capability and always returns `None`.

## Suggested Fix

Update `FAISSIndex.get_vector()` to reconstruct vectors whenever the underlying index supports it.

Suggested approach:

1. Convert the string `vector_id` into its corresponding integer position using `self.vector_ids`.
2. Call:

```python
self.index.reconstruct(idx)
```

3. Return the reconstructed NumPy array.
4. If reconstruction is unsupported by the current FAISS index type, gracefully return `None` instead of failing.

## Benefits

- Enables retrieval of stored embeddings.
- Restores feature parity with the in-memory backend.
- Avoids silently discarding functionality that `IndexFlatL2` already provides.

### PR DESCRIPTION
**Pull request by SaurabhScripts**
## Description

Enable `FAISSIndex.get_vector()` to return stored embeddings when the underlying FAISS index supports reconstruction. The method maps the public vector ID to its FAISS position, calls `reconstruct()`, and returns a `float32` NumPy array. Missing IDs and indexes with known unsupported-reconstruction errors return `None`; unexpected FAISS runtime failures remain visible to callers.

## Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Performance improvement
- [ ] Code refactoring

## Related Issues

Fixes #844

## Changes Made

- Resolve stored vector IDs to their FAISS positional index.
- Reconstruct vectors through the backend when supported.
- Return `None` for missing IDs, `NotImplementedError`, and known FAISS unsupported-reconstruction messages.
- Re-raise unexpected `RuntimeError` failures so broken index state remains diagnosable.
- Add unit coverage plus a real `faiss.IndexFlatL2` reconstruction test.

## Testing

- [x] Tested locally
- [x] Added tests for new functionality
- [x] Package builds successfully (`python -m build`)

### Test Commands

```bash
python -m pytest tests/vector_store/test_faiss_index.py -q
# 8 passed

python -m pytest tests/vector_store -m "not integration" \
  --ignore=tests/vector_store/test_pinecone_store.py -q
# 134 passed, 61 skipped, 11 deselected

python -m black --check semantica/vector_store/faiss_store.py \
  tests/vector_store/test_faiss_index.py
python -m isort --check-only semantica/vector_store/faiss_store.py \
  tests/vector_store/test_faiss_index.py
git diff --check
python -m build
# Successfully built sdist and wheel
```

The complete non-integration vector-store run was also attempted. Six unrelated Pinecone tests require the optional Pinecone client and fail their availability guard when it is absent; all other collected vector-store tests pass.

## Documentation

- [ ] Updated relevant documentation
- [ ] Added code examples if applicable
- [ ] Updated API reference if adding new APIs
- [ ] Updated cookbook if adding new examples
- [x] No documentation changes needed

## Breaking Changes

**Breaking Changes**: No

## Checklist

- [x] My code follows the project's style guidelines
- [x] I have performed a self-review of my code
- [x] I have commented my code where the logic is not self-explanatory
- [x] My changes generate no new warnings
- [x] Package builds successfully

## Additional Notes

Unsupported reconstruction matching is intentionally limited to FAISS messages containing `reconstruct not implemented` or `reconstruct_from_offset not implemented`; other runtime errors are re-raised.

### CODE REVIEW

File: semantica/vector_store/faiss_store.py [hunk 1]
```diff
@@ -85,9 +85,14 @@ def get_vector(self, vector_id: str) -> Optional[np.ndarray]:
...
-        # This is a simplified approach
-        return None
+        reconstruct = getattr(self.index, "reconstruct", None)
+        if not callable(reconstruct):
+            return None
+
+        try:
+            return np.asarray(reconstruct(idx), dtype=np.float32)
+        except (NotImplementedError, RuntimeError):
+            return None
```
[the remark this answers is not in the document]
Comment (SaurabhScripts) [reply semantica/vector_store/faiss_store.py:95]: Addressed in `d2f15f4`. `NotImplementedError` and known FAISS unsupported-reconstruction messages still return `None`, while unexpected `RuntimeError` failures such as `direct map not initialized` are re-raised unchanged. Regression tests cover both unsupported variants and unexpected runtime propagation. All 8 focused tests and 134 non-integration vector-store tests pass.

**Comment by Sameer6305** [comment]
@KaifAhmad1 
I reviewed the implementation and found one edge case I'd like to clarify before suggesting a code change.

IndexIVFFlat without an initialized DirectMap raises RuntimeError: direct map not initialized. The current implementation propagates this error, whereas other unsupported reconstruction cases return None.

Should this DirectMap case also be treated as "reconstruction unavailable" (return None), or is propagating the runtime error the intended API behavior?

I have a small implementation and regression test ready if the former behavior is preferred.

**Comment by SaurabhScripts** [comment]
Current status: all checks pass, but this branch now conflicts with `main`, which contains overlapping FAISS retrieval changes. The remaining behavior question is whether `IndexIVFFlat` with an uninitialized DirectMap should return `None` as reconstruction-unavailable or propagate `RuntimeError`. I can update the branch and resolve the overlap once that API behavior is confirmed.

**Comment by KaifAhmad1** [comment]
@Sameer6305 @SaurabhScripts — merged `main` in and resolved the conflict.

**On the DirectMap question**: neither "return None" nor "propagate" was quite right. `IndexIVFFlat` *can* reconstruct exact vectors, it just needs `make_direct_map()` called once first — that's different from indices that can never reconstruct. Since this repo's `FAISSIndexBuilder.build_index(index_type="ivf")` builds exactly that index type and nothing ever called `make_direct_map()`, every IVF user would have hit this error every time. So `get_vector()` now catches `"direct map not initialized"` specifically, builds the direct map lazily, and retries — reconstruction actually works for IVF now instead of failing or silently returning `None`.

Also resolved the conflict with `main`'s `except Exception: return None` (landed via #847 while this was open) in favor of this branch's more precise `NotImplementedError`/`RuntimeError`-message handling, now extended with the DirectMap case.

Added 3 tests: `make_direct_map` retry via mock, `None` fallback when it's unavailable, and a real `faiss.IndexIVFFlat` end-to-end reconstruction test. Full suite: 168 passed, 0 failed. Pushed to this branch — should be conflict-free now, ready for your review @Sameer6305.

**Review by KaifAhmad1 — APPROVED** [review]
Reviewed the final diff after the merge/DirectMap fix.

- `get_vector()` correctly distinguishes three cases: permanently-unsupported reconstruction (`NotImplementedError` / known "not implemented" messages → `None`), the fixable `IndexIVFFlat` DirectMap gap (lazily built + retried), and genuinely unexpected `RuntimeError`s (re-raised so broken index state stays diagnosable).
- Test coverage is solid: mocked edge cases plus real `faiss.IndexFlatL2` and `faiss.IndexIVFFlat` reconstruction tests.
- All CI checks green, no merge conflicts.

Approving. Closes #844.
